### PR TITLE
Add kyuubi-shaded module to avoid conflicts with external jars

### PR DESCRIPTION
**Pull request by turboFei**
turboFei [Closes #488](https://github.com/yaooqinn/kyuubi/pull/488)    Target Issue Feature [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### _Why are the changes needed?_
Shade curator dependency in kyuubi-ha module to avoid impaction by external jars.

### _How was this patch tested?_
Existing UT.

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -105,6 +105,8 @@
         <maven.plugin.download.version>1.6.0</maven.plugin.download.version>
         <maven.plugin.silencer.version>1.6.0</maven.plugin.silencer.version>
         <maven.plugin.rat.version>0.13</maven.plugin.rat.version>
+        <!-- Package to use when relocating shaded classes. -->
+        <kyuubi.shade.packageName>org.kyuubi_project</kyuubi.shade.packageName>
```
Comment (yaooqinn) [inline pom.xml:109]: `org.apache.kyuubi.shade` looks more clear to me

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 2]
```diff
@@ -167,8 +167,18 @@
                         <includes>
                             <include>org.apache.kyuubi:kyuubi-common</include>
                             <include>org.apache.kyuubi:kyuubi-ha</include>
+                            <include>org.apache.curator:*</include>
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/pom.xml:170]: can we avoid using wildcard here?

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 3]
```diff
@@ -167,8 +167,18 @@
...
                         </includes>
                     </artifactSet>
+                    <relocations>
+                        <relocation>
+                            <pattern>org.apache.curator</pattern>
+                            <shadedPattern>${kyuubi.shade.packageName}.curator</shadedPattern>
+                            <includes>
+                                <include>org.apache.curator.**</include>
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/pom.xml:178]: what about the transitive dependencies, shall we explicitly exclude them or add them here too? cc @pan3793

File: externals/kyuubi-spark-sql-engine/pom.xml (hunk 2 above)
Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/pom.xml]: I tried to include curator-framework and curator-recipes here.

Let me wait the ut result and determine whether also include curator-test here.

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/pom.xml]: ok，include curator-client

**Comment by pan3793** [comment]
Do shade on `kyuubi-spark-sql-engine` looks like not safe for me.
If we define a different version of curator and hive from spark, it will overwrite transitive dependencies of spark in `kyuubi-spark-sql-engine` compile and test scopes. 

For example, kyuubi uses curator4 and hive3, but spark30 depends curator2 and hive2.
In `kyuubi-spark-sql-engine` compile and test scopes, the actually effected version is curator4 and hive3.

**Comment by pan3793** [comment]
I recommend to pre-shade the `hive-services-rpc` and `curator-*` in separated modules, and let `kyuubi-spark-engine` depends it. (or we should let `kyuubi-common` depends it, and other modules just use the transitive dependencies).

AFAK, Apache Flink use this approach, and they maintains the pre-shaded dependencies in a separated repo https://github.com/apache/flink-shaded . Apache Hadoop has a plan to migrate such approach like Flink.

**Comment by pan3793** [comment]
If we chose the pre-shade way, we need change `import` in codebase to reference the shaded versions.

**Comment by turboFei** [comment]
thanks for the proposal, it sounds good.

**Comment by yaooqinn** [comment]
Does pre-shade way has any adverse impact, such as users want another curator client?

**Comment by pan3793** [comment]
> Does pre-shade way has any adverse impact, such as users want another curator client?

If user want to extend kyuubi original functions, i.e. custom the auth of curator, the should use kyuubi shaded curator api.

If user want to add individual functions to kyuubi, i.e. integration kyuubi to corp micro services discovery framework, and it happens to use curator, in such scenario normal curator will co-exist will kyuubi shaded curator.

**Comment by yaooqinn** [comment]
SGTM

**Comment by pan3793** [comment]
Apache Drill https://github.com/apache/drill pre-shade guava in a module, let's follow it.

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 4]
```diff
@@ -98,6 +98,12 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-service-rpc</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (turboFei) [inline externals/kyuubi-spark-sql-engine/pom.xml:105]: we need it here, because it could provide org.apache.thrift.classes for the test.

**Comment by turboFei** [comment]
ci passed finally @pan3793

File: .travis.yml [hunk 5]
```diff
@@ -47,6 +47,9 @@ matrix:
 
 install:
   - mvn --version
+  - cd kyuubi-shaded
+  - mvn clean install
+  - cd -
```
Comment (pan3793) [inline .travis.yml:52]: Can it be replaced by `mvn clean install -pl <modules> -am -Pkyuubi-shaded`?

File: kyuubi-shaded/kyuubi-shaded-curator/pom.xml [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    <parent>
+        <artifactId>kyuubi-shaded</artifactId>
+        <groupId>org.apache.kyuubi</groupId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>kyuubi-shaded-curator</artifactId>
+    <name>Kyuubi Project Shaded curator</name>
```
Comment (yaooqinn) [inline kyuubi-shaded/kyuubi-shaded-curator/pom.xml:31]: Kyuubi Project Shade Apache Curator

File: kyuubi-shaded/kyuubi-shaded-curator/pom.xml [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                <artifactId>maven-shade-plugin</artifactId>
+                <configuration>
+                    <shadedArtifactAttached>false</shadedArtifactAttached>
+                    <artifactSet>
+                        <includes>
+                            <include>org.apache.curator:curator-client</include>
+                            <include>org.apache.curator:curator-framework</include>
+                            <include>org.apache.curator:curator-recipes</include>
+                            <include>org.apache.curator:curator-test</include>
+                            <include>org.javassist:javassist</include>
```
Comment (yaooqinn) [inline kyuubi-shaded/kyuubi-shaded-curator/pom.xml:112]: is there any transitive we need?

File: kyuubi-shaded/kyuubi-shaded-curator/pom.xml [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                        </includes>
+                    </artifactSet>
+                    <relocations>
+                        <relocation>
+                            <pattern>org.apache.curator</pattern>
+                            <shadedPattern>${kyuubi.shade.packageName}.curator</shadedPattern>
```
Comment (yaooqinn) [inline kyuubi-shaded/kyuubi-shaded-curator/pom.xml:118]: `${kyuubi.shade.packageName}.org.apache.curator` looks more future-proofing, as maybe there is another project called `curator` :)

File: kyuubi-shaded/kyuubi-shaded-curator/pom.xml [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                            <includes>
+                                <include>org.apache.curator.**</include>
+                            </includes>
+                        </relocation>
+                        <relocation>
+                            <pattern>org.javassist</pattern>
+                            <shadedPattern>${kyuubi.shade.packageName}.javassist</shadedPattern>
```
Comment (yaooqinn) [inline kyuubi-shaded/kyuubi-shaded-curator/pom.xml:125]: ditto

File: kyuubi-shaded/kyuubi-shaded-hive-service-rpc/pom.xml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    <parent>
+        <artifactId>kyuubi-shaded</artifactId>
+        <groupId>org.apache.kyuubi</groupId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>kyuubi-shaded-hive-service-rpc</artifactId>
+    <name>Kyuubi Project Shaded Hive Service Rpc</name>
```
Comment (yaooqinn) [inline kyuubi-shaded/kyuubi-shaded-hive-service-rpc/pom.xml:31]: Kyuubi Project Shade Hive Service RPC

**Comment by pan3793** [comment]
https://github.com/yaooqinn/kyuubi/pull/488/files#L1038

As we have shade module now, it's not suitable at parent `pom.xml`

File: .travis.yml (hunk 5 above)
Comment (pan3793) [reply .travis.yml:52]: `mvn clean install -pl :kyuubi-shaded-curator,:kyuubi-shaded-hive-service-rpc -am -Pkyuubi-shaded`

File: pom.xml [hunk 11]
```diff
@@ -1440,5 +1376,11 @@
                 <spark.archive.download.skip>true</spark.archive.download.skip>
             </properties>
         </profile>
+        <profile>
+            <id>kyuubi-shaded</id>
+            <modules>
+                <module>kyuubi-shaded</module>
```
Comment (yaooqinn) [inline pom.xml:1382]: why it is a profile?

File: .github/workflows/master.yml [hunk 12]
```diff
@@ -67,6 +67,8 @@ jobs:
           key: ${{ runner.os }}-maven-io-${{ hashFiles('**/pom.xml') }}
           restore-keys: |
             ${{ runner.os }}-maven-io-
+      - name: install kyuubi-shaded jars
+        run: cd kyuubi-shaded && mvn clean install && cd -
```
Comment (yaooqinn) [inline .github/workflows/master.yml:71]: can it not be installed in a single `mvn install`?

File: pom.xml (hunk 11 above)
Comment (pan3793) [reply pom.xml:1382]: After some investigation with @turboFei, we found that maven has some limitation of shade module reference, when run `mvn package`, the downstream module can't see shaded classes of upstream module which in same maven project.

There is a solution https://stackoverflow.com/a/50274253 but introduced another issue, we must run `mvn clean xxx` everytime, w/o `mvn clean` will cause compile failed.

Maybe that's why Apache Flink and Apache Hadoop choose shade deps in separated module, and Apache Iceberg which use Gradle doesn't suffer this issue.

Comment (yaooqinn) [reply pom.xml:1382]: So this do related to the releasing process and a doc is also needed

Comment (yaooqinn) [reply pom.xml:1382]: So this is related to the releasing process, and a documentation is also needed

File: kyuubi-shaded/kyuubi-shaded-curator/pom.xml (hunk 7 above)
Comment (turboFei) [reply kyuubi-shaded/kyuubi-shaded-curator/pom.xml:112]: we need javassist, otherwise exception would be thrown.

File: pom.xml (hunk 11 above)
Comment (turboFei) [reply pom.xml:1381]: is it ok to add some notes into building doc?

Comment (turboFei) [reply pom.xml:1381]: https://github.com/yaooqinn/kyuubi/pull/488/files?file-filters%5B%5D=.md&file-filters%5B%5D=.scala&file-filters%5B%5D=.xml#diff-434e619c9343e148d9ef7e85b044633eb86f72c6f1137eac287a97708f18c86cR11

**Comment by turboFei** [comment]
after remove
```
    <parent>
        <artifactId>kyuubi</artifactId>
        <groupId>org.apache.kyuubi</groupId>
        <version>1.2.0-SNAPSHOT</version>
        <relativePath>../pom.xml</relativePath>
    </parent>
```
in kyuubi-shaded module, there is no transaction dependencies.
https://github.com/yaooqinn/kyuubi/pull/488/commits/224cc38b45d5a02acd9fcc45d2902b60113f2abf

**Review by pan3793 — APPROVED** [review]
LGTM, thanks!

**Comment by pan3793** [comment]
There are some things we need consider after switch to shaded.

- License declaration.
- Add rule to force use shaded version in Kyuubi codebase.

And we can do those in following PRs.

**Comment by pan3793** [comment]
After this change, `kyuubi-spark-sql-engine` expect to work with spark build by without `-Phive-thriftserver`, right? @turboFei

**Comment by yaooqinn** [comment]
Can we have a module like `test/e2e` to verify `./build/dist --tgz` still works in followups?

**Comment by yaooqinn** [comment]
> After this change, `kyuubi-spark-sql-engine` is expected to work with spark build without `-Phive-thriftserver`, right? @turboFei

Hmm? Do you mean that we are going to drop support for official spark releases?

**Comment by pan3793** [comment]
> > After this change, `kyuubi-spark-sql-engine` is expected to work with spark build without `-Phive-thriftserver`, right? @turboFei
> 
> Hmm? Do you mean that we are going to drop support for official spark releases?

Kyuubi always works with spark official release. And if we use custom spark dist, I want to minimize the deps.

**Comment by turboFei** [comment]
> > After this change, `kyuubi-spark-sql-engine` is expected to work with spark build without `-Phive-thriftserver`, right? @turboFei
> 
> Hmm? Do you mean that we are going to drop support for official spark releases?

I think he means that we can support the spark release without hive-thritfserver after this change.

**Comment by turboFei** [comment]
> Can we have a module like `test/e2e` to verify `./build/dist --tgz` still works in followups?

sure, will try

**Comment by yaooqinn** [comment]
I see. I misunderstood. LGTM

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
note that, we have an auto pipeline for GA release, we also need to verify the scripts and all the steps are still well

**Comment by turboFei** [comment]
ok, I will update ./build/dist to add install shaded modules part.
and verify it in ci/GA

**Comment by yaooqinn** [comment]
thanks, merged to master for v1.2.0

**Comment by yaooqinn** [comment]
>mvn clean install -pl :kyuubi-shaded-curator,:kyuubi-shaded-hive-service-rpc -am -Pkyuubi-shaded

is this warning expected?

[WARNING] The requested profile "kyuubi-shaded" could not be activated because it does not exist.

**Comment by pan3793** [comment]
> > mvn clean install -pl :kyuubi-shaded-curator,:kyuubi-shaded-hive-service-rpc -am -Pkyuubi-shaded
> 
> is this warning expected?
> 
> [WARNING] The requested profile "kyuubi-shaded" could not be activated because it does not exist.

Add to `kyuubi-shaded/pom.xml` will suppress this warning, but without it impact nothing.
```
   <profiles>
        <profile>
            <id>kyuubi-shaded</id>
        </profile>
    </profiles>
```

**Comment by pan3793** [comment]
Although we have reached a consensus that pre-shade is a right direction, but such build step confuse developers new to Kyuubi, so I temporarily revert it and let us do pre-shade in separated repo or find some way more friendly for developers.

**Comment by yaooqinn** [comment]
Let keep the old school way by shading curator to the engine module as a46f00f396c8191a3e45254b50fcf8979372c909, thanks @pan3793, and sorry for the back and forth @turboFei

**Pull request closed without merging**